nvflare/app_common/resource_managers/list_resource_manager.py

```diff
@@ -12,66 +12,138 @@
+        if expiration_period < 0:
+            raise ValueError("expiration_period should be greater than 0.")
+
         self.resources = {}
         for k in resources:
             if not isinstance(resources[k], list):
                 raise TypeError(f"item in resources should be of type list, but got {type(resources[k])}.")
-            self.resources[k] = set(resources[k])
+            self.resources[k] = deque(list(dict.fromkeys(resources[k])))
+
+        self.expiration_period = expiration_period
+        self.reserved_resources = {}
+        self.lock = Lock()
+        self.stop_event = Event()
+        self.cleanup_thread = Thread(target=self._check_expired)
+
+    def _check_expired(self):
+        while not self.stop_event.is_set():
+            time.sleep(1)
+            with self.lock:
+                tokens_to_remove = []
+                for k in self.reserved_resources:
+                    r, t = self.reserved_resources[k]
+                    t -= 1
+                    if t == 0:
+                        tokens_to_remove.append(k)
+                    else:
+                        self.reserved_resources[k] = r, t
+                for token in tokens_to_remove:
+                    reserved_resources, _ = self.reserved_resources.pop(token)
+                    for k in reserved_resources:
+                        for i in reserved_resources[k]:
+                            self.resources[k].append(i)
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.SYSTEM_START:
+            self.cleanup_thread.start()
+        elif event_type == EventType.SYSTEM_END:
+            self.stop_event.set()
+            if self.cleanup_thread:
+                self.cleanup_thread.join()
+                self.cleanup_thread = None
 
     def check_resources(self, resource_requirement: dict, fl_ctx: FLContext) -> (bool, Optional[str]):
         if not isinstance(resource_requirement, dict):
             raise TypeError(f"resource_requirement should be of type dict, but got {type(resource_requirement)}.")
+
         check_result = True
-        for k in resource_requirement:
-            if k in self.resources:
-                if len(self.resources[k]) < resource_requirement[k]:
+        token = None
+        with self.lock:
+            for k in resource_requirement:
+                if k in self.resources:
+                    if len(self.resources[k]) < resource_requirement[k]:
+                        check_result = False
+                        self.log_debug(fl_ctx, f"Resource {k} is not enough.")
+                        break
+                else:
                     check_result = False
+                    self.log_debug(fl_ctx, f"Missing {k} in resources.")
                     break
-        return check_result, str(uuid.uuid4())
+
+        # reserve resource only when check is True
+        if check_result:
```

Comment:
same comment here: 
    if we assume cpu resource is available, but gpu resource is NOT available,  the check_result is still TRUE. 
     The following logic will allocate the resources for both cpu and gpu.

Reply:
For the current logic, if CPU resource is available but GPU is NOT available, then "check_result" will be False.